src/NonLinearSolver.f90

```diff
@@ -0,0 +1,417 @@
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!                          Futility Development Group                          !
+!                             All rights reserved.                             !
+!                                                                              !
+! Futility is a jointly-maintained, open-source project between the University !
+! of Michigan and Oak Ridge National Laboratory.  The copyright and license    !
+! can be found in LICENSE.txt in the head directory of this repository.        !
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!> @brief Module provides a non-linear solver class
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+MODULE NonLinearSolverModule
+#include "Futility_DBC.h"
+USE Futility_DBC
+USE ISO_C_BINDING
+USE IntrType
+USE Constants_Conversion
+USE BLAS
+USE ParameterLists
+USE ParallelEnv
+USE FutilityComputingEnvironmentModule
+USE VectorTypes
+USE MatrixTypes
+USE LinearSolverTypes
+
+IMPLICIT NONE
+PRIVATE
+
+PUBLIC :: NonLinearSolverInterface_Base
+PUBLIC :: NonLinearSolver_Base
+PUBLIC :: NonLinearSolver_Native
+#ifdef UNIT_TEST
+PUBLIC :: testSolverMethod
+PUBLIC :: testTPLType
+PUBLIC :: testN
+PUBLIC :: testIterations
+PUBLIC :: testTol
+PUBLIC :: testLinSys
+PUBLIC :: testFunc
+PUBLIC :: testCE
+#endif
+
+!> Non-linear solver TPL enumerations
+INTEGER(SIK),PARAMETER,PUBLIC :: NLSOLVER_TPL_NATIVE=1
+!> Non-linear solver method enumerations
+INTEGER(SIK),PARAMETER,PUBLIC :: NLSOLVER_METHOD_BISECTION=1
+INTEGER(SIK),PARAMETER,PUBLIC :: NLSOLVER_METHOD_NEWTON=2
+INTEGER(SIK),PARAMETER :: VALID_NLSOLVER_METHODS(2)=(/NLSOLVER_METHOD_BISECTION, &
+    NLSOLVER_METHOD_NEWTON/)
+
+!> Abstract type for the non-linear solver interface.  Defines procedures
+!> for evaluating the function and its jacobian.  The client should extend
+!> this class for its specific applications.
+TYPE,ABSTRACT :: NonLinearSolverInterface_Base
+  PRIVATE
+  CONTAINS
+    !> @copybrief NonLinearSolverModule::clearInterface_Base
+    !> @copydetails NonLinearSolverModule::clearInterface_Base
+    PROCEDURE,PASS :: clear => clearInterface_Base
+    !> @copybrief NonLinearSolverModule::nonlinearsolver_eval_absintfc
+    !> @copydetails NonLinearSolverModule::nonlinearsolver_eval_absintfc
+    PROCEDURE(nonlinearsolver_eval_absintfc),DEFERRED,PASS :: eval
+    !> @copybrief NonLinearSolverModule::nonlinearsolver_jacobian_absintfc
+    !> @copydetails NonLinearSolverModule::nonlinearsolver_jacobian_absintfc
+    PROCEDURE(nonlinearsolver_jacobian_absintfc),DEFERRED,PASS :: jacobian
```

Comment:
We had discussed writing a base jacobian that estimated it.

Reply:
Moved from Nodal to the base type